File: examples/C/atomic_exercise.c
#include <stdio.h>
#include <stdlib.h>
int main(void)
{
    // Histogram allocation and initialization
    int histo[10];
    for (int i=0; i<10; ++i)
        histo[i] = 0;
    size_t nshots = (size_t) 1e9;
    
    // Allocate memory for the random numbers
    int* shots = (int*) malloc(nshots*sizeof(int));

    srand((unsigned) 12345900);     
    
    // Fill the array on the CPU (rand is not available on GPU with Nvidia Compilers)
    for (size_t i=0; i< nshots; ++i)
    {
        shots[i] = (int) rand() % 10;
    }
    
    // Count the number of time each number was drawn 
    for (size_t i=0; i<nshots; ++i)
    {
        histo[shots[i]]++;
    }
    
    // Print results
    
    for (int i=0; i<10; ++i)
        printf("%3d: %10d (%5.3f)\n", i, histo[i], (double) histo[i]/1.e9);
      
    return 0;
}

File: examples/Fortran/atomic_exercise.f90
program histogram
    use iso_fortran_env, only : REAL64, INT32
    implicit none

    integer(kind=INT32 ), dimension(:) , allocatable :: shots
    integer(kind=INT32 ), dimension(10)              :: histo
    integer(kind=INT32 ), parameter                  :: nshots = 1e9
    real   (kind=REAL64)                             :: random_real
    integer(kind=INT32 )                             :: i

    ! Histogram allocation and initialization
    do i = 1, 10
     histo(i) = 0
    enddo

    ! Allocate memory for the random numbers
    allocate(shots(nshots))

    ! Fill the array on the CPU (rand is not available on GPU with Nvidia Compilers)
    do i = 1, nshots
        call random_number(random_real)
        shots(i) = floor(random_real * 10.0_real64) + 1
    enddo

    ! Count the number of time each number was drawn
    do i = 1, nshots
        histo(shots(i)) = histo(shots(i)) + 1
    enddo

    !  Print results
    do i = 1, 10
        write(0,"(i2,a2,i10,a2,f10.8,a1)") i,": ", histo(i), " (", real(histo(i))/1.e9, ")"
    enddo

    deallocate(shots)

end program histogram
